### ISSUE TITLE
test client with json body needs conversion to dict() not json()?

### ISSUE THREAD

**Issue by gwdekker** [tiangolo/fastapi#192]
Thanks for the fantastic library, and the excellent documentation!

I have a POST method ```/{model_name}/predict``` that takes a pydantic object as body looking like this:

```python
class PredictRequest(BaseModel):
    columns: List[str]
    index: List[int]
    data: List[List[Optional[float]]]
```
I was expecting to be able to test this using the following (where client is a pytest fixture creating the starlette testclient): 

```python
def test_predict_post(client):
    area = "NO1"
    predict_request = PredictRequest(
        columns=["test"], index=[1, 2], data=[[3], [4]]
    ).json()
    response = client.post(f"/{area}/predict", json=predict_request)
    print(response.json())
    assert response.status_code == HTTP_200_OK
```
When running this i got an error message like this:

```python
{'detail': [{'loc': ['body', 'predict_input'], 'msg': 'value is not a valid dict', 'type': 'type_error.dict'}]}
```

The solution is to instead convert to dict not json:
```python
def test_predict_post(client):
    area = "NO1"
    predict_request = PredictRequest(
        columns=["test"], index=[1, 2], data=[[3], [4]]
    ).dict()
    response = client.post(f"/{area}/predict", json=predict_request)
    print(response.json())
    assert response.status_code == HTTP_200_OK
```

Is this the expected behaviour? It seemed counter-intuitive to me and took me a while to figure out so decided to document it here for others to use (or for the dev team to consider changing or documenting on the test documentation).

**Comment by euri10** [comment]
Not a bug, iirc the json kwarg from the test client expects a dictionary,
much like in the requests library, this is part of Starlette btw, not
FastAPI.
The dict() method of pydantic allows you to do that, it's json() method
returns a string however.

Le lun. 29 avr. 2019 à 8:45 AM, gwdekker <notifications@github.com> a
écrit :

**Comment by gwdekker** [comment]
Yes, thanks, the bug label should probably be dropped. I would suggest though to have this in the test documentation as it is a bit counterintuitive when working with this for the first time.

**Comment by euri10** [comment]
It happened to me even if accustomed with requests, so the addition would
be helpful.

While I read your test, something additional comes to mind : I prefer to
use something like client.post(url_for) instead of a hard coded f-string or
string for the url, this enables copy paste of blueprints among
applications you may write and use same tests case should your API prefixes
vary.
Sorry for no formatting, afk!


Le lun. 29 avr. 2019 à 9:00 AM, gwdekker <notifications@github.com> a
écrit :

**Comment by tiangolo** [comment]
Thanks for your help here @euri10 !

Thanks for the report @gwdekker, I'll have it in mind to add a more explicit example in the docs.

**Comment by tiangolo** [comment]
@gwdekker I updated the docs here with more info covering a bit more those cases: https://fastapi.tiangolo.com/tutorial/testing/#testing-extended-example

You might still find useful some of the information in @euri10's PR here: https://github.com/tiangolo/fastapi/pull/205

**Comment by gwdekker** [comment]
Great extension! Thanks for taking the time to look into this!

**Comment by tiangolo** [comment]
Great! I'm glad that helped. I'll close this issue now then, but feel free to add more comments or create new issues.
